File: 3625.count-number-of-trapezoids-ii.go
#
# @lc app=leetcode id=3625 lang=golang
#
# [3625] Count Number of Trapezoids II
#

# @lc code=start
import "fmt"

func countTrapezoids(points [][]int) int {
	n := len(points)
	if n < 4 {
		return 0
	}

	const OFF = 3000
	const MS = 6001
	type Key uint32

	indexmap := make(map[Key]int)
	for i := 0; i < n; i++ {
		x := points[i][0]
		y := points[i][1]
		k := Key(uint32(x + OFF) + MS * uint32(y + OFF))
		indexmap[k] = i
	}

	// Compute number of parallelograms * 8
	var para int64 = 0
	for i := 0; i < n; i++ {
		pix := points[i][0]
		piy := points[i][1]
		for j := 0; j < n; j++ {
			if j == i {
				continue
			}
			pjx := points[j][0]
			pjy := points[j][1]
			dx1 := pjx - pix
			dy1 := pjy - piy
			for k := 0; k < n; k++ {
				if k == i || k == j {
					continue
				}
				pkx := points[k][0]
				pky := points[k][1]
				dx2 := pkx - pix
				dy2 := pky - piy
				cross := int64(dx1) * int64(dy2) - int64(dy1) * int64(dx2)
				if cross == 0 {
					continue
				}
				px := pjx + pkx - pix
				py := pjy + pky - piy
				ptk := Key(uint32(px + OFF) + MS * uint32(py + OFF))
				idx, ok := indexmap[ptk]
				if !ok || idx == i || idx == j || idx == k {
					continue
				}
				para++
			}
		}
	}

	P := para / 8

	// Slope constants
	const DYOFF = 2000
	const DYMAX = 4001
	type SlopeKey int64

	numEdges := make(map[SlopeKey]int)

	abs := func(a int) int {
		if a < 0 {
			return -a
		}
		return a
	}

	gcd := func(a, b int) int {
		a = abs(a)
		b = abs(b)
		for b != 0 {
			a, b = b, a % b
		}
		return a
	}

	normalize := func(dx, dy int) (int, int) {
		g := gcd(dx, dy)
		if g == 0 {
			return 0, 0
		}
		dx /= g
		dy /= g
		if dx < 0 || (dx == 0 && dy < 0) {
			dx = -dx
			dy = -dy
		}
		return dx, dy
	}

	// Compute numEdges per slope
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			dx := points[j][0] - points[i][0]
			dy := points[j][1] - points[i][1]
			ndx, ndy := normalize(dx, dy)
			sk := SlopeKey(int64(ndx) * int64(DYMAX) + int64(ndy + DYOFF))
			numEdges[sk]++
		}
	}

	// Compute S
	var S int64 = 0
	for sk, totseg := range numEdges {
		if totseg < 2 {
			continue
		}
		ndx := int(sk / DYMAX)
		ndy := int(sk % DYMAX) - DYOFF
		linecnt := make(map[int64]int)
		for _, p := range points {
			c := int64(ndy) * int64(p[0]) - int64(ndx) * int64(p[1])
			linecnt[c]++
		}
		var sumsq int64 = 0
		for _, cnt := range linecnt {
			if cnt < 2 {
				continue
			}
			segs := int64(cnt) * int64(cnt-1) / 2
			sumsq += segs * segs
		}
		tot64 := int64(totseg)
		contrib := (tot64 * tot64 - sumsq) / 2
		S += contrib
	}

	ans := S - P
	if ans < 0 {
		return 0 // safety
	}
	return int(ans)
}

# @lc code=end

File: 3625.count-number-of-trapezoids-ii.java
#
# @lc app=leetcode id=3625 lang=java
#
# [3625] Count Number of Trapezoids II
#

# @lc code=start
class Solution {
    private static class Slope {
        int dx, dy;

        Slope(int ddx, int ddy) {
            int adx = Math.abs(ddx);
            int ady = Math.abs(ddy);
            int g = gcd(adx, ady);
            dx = ddx / g;
            dy = ddy / g;
            if (dx < 0 || (dx == 0 && dy < 0)) {
                dx = -dx;
                dy = -dy;
            }
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            Slope other = (Slope) obj;
            return dx == other.dx && dy == other.dy;
        }

        @Override
        public int hashCode() {
            return 31 * dx + dy;
        }
    }

    private static int gcd(int a, int b) {
        while (b != 0) {
            int t = b;
            b = a % b;
            a = t;
        }
        return a;
    }

    private static long getKey(int x, int y) {
        final int OFFSET = 1005;
        final long MUL = 2005L;
        return ((long) x + OFFSET) * MUL + (y + OFFSET);
    }

    public int countTrapezoids(int[][] points) {
        int n = points.length;
        java.util.Set<Slope> slopeSet = new java.util.HashSet<>();
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                int dxp = points[j][0] - points[i][0];
                int dyp = points[j][1] - points[i][1];
                slopeSet.add(new Slope(dxp, dyp));
            }
        }

        long total = 0;
        for (Slope s : slopeSet) {
            int dx = s.dx;
            int dy = s.dy;
            java.util.Map<Long, Integer> lineCnt = new java.util.HashMap<>();
            for (int p = 0; p < n; ++p) {
                long lkey = (long) dx * points[p][1] - (long) dy * points[p][0];
                lineCnt.merge(lkey, 1, Integer::sum);
            }
            long sumC = 0;
            long sumCsq = 0;
            for (int c : lineCnt.values()) {
                if (c < 2) continue;
                long cc = (long) c * (c - 1) / 2;
                sumC += cc;
                sumCsq += cc * cc;
            }
            long contrib = (sumC * sumC - sumCsq) / 2;
            total += contrib;
        }

        // Count parallelograms
        java.util.Map<Long, Integer> ptIdx = new java.util.HashMap<>();
        for (int i = 0; i < n; ++i) {
            long key = getKey(points[i][0], points[i][1]);
            ptIdx.put(key, i);
        }

        long paraTimes = 0;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (j == i) continue;
                long vx = points[j][0] - points[i][0];
                long vy = points[j][1] - points[i][1];
                for (int k = 0; k < n; ++k) {
                    if (k == i || k == j) continue;
                    long wx = points[k][0] - points[i][0];
                    long wy = points[k][1] - points[i][1];
                    long cross = vx * wy - vy * wx;
                    if (cross == 0) continue;
                    int dx = points[j][0] + points[k][0] - points[i][0];
                    int dy_ = points[j][1] + points[k][1] - points[i][1];
                    long dkey = getKey(dx, dy_);
                    Integer lobj = ptIdx.get(dkey);
                    if (lobj != null) {
                        int l = lobj;
                        if (l != i && l != j && l != k) {
                            ++paraTimes;
                        }
                    }
                }
            }
        }
        long numPara = paraTimes / 8;

        return (int) (total - numPara);
    }
}
# @lc code=end